text/0000-global-allocators.md

```diff
@@ -0,0 +1,248 @@
+developers should be making that decision, not the compiler. The system
+allocator is a more reasonable and unsurprising default choice.
+
+A third party crate allowing users to opt-into jemalloc would also open the door
+to provide access to some of the library's other features such as tracing, arena
+pinning, and diagnostic output dumps for code that depends on jemalloc directly.
+
+[this comment]: https://github.com/rust-lang/rust/issues/36963#issuecomment-252029017
+
+# Detailed design
+[design]: #detailed-design
+
+## Defining an allocator
+
+An allocator crate identifies itself as such by applying the `#![allocator]`
+annotate at the crate root. It then defines a specific set of functions which
+are tagged with attributes:
+
+```rust
+#![allocator]
+
+/// Returns a pointer to `size` bytes of memory aligned to `align`.
+///
+/// On failure, returns a null pointer.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+#[allocator(allocate)]
+pub fn allocate(size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Returns a pointer to `size` bytes of memory aligned to `align`, and
+/// initialized with zeroes.
+///
+/// On failure, returns a null pointer.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+#[allocator(allocate_zeroed)]
+pub fn allocate_zeroed(size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Deallocates the memory referenced by `ptr`.
+///
+/// The `ptr` parameter must not be null.
+///
+/// The `old_size` and `align` parameters are the parameters that were used to
+/// create the allocation referenced by `ptr`.
+#[allocator(deallocate)]
+pub fn deallocate(ptr: *mut u8, old_size: usize, align: usize) {
+    ...
+}
+
+/// Resizes the allocation referenced by `ptr` to `size` bytes.
+///
+/// On failure, returns a null pointer and leaves the original allocation
+/// intact.
+///
+/// If the allocation was relocated, the memory at the passed-in pointer is
+/// undefined after the call.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
+/// power of 2. The alignment must be no larger than the largest supported page
+/// size on the platform.
+///
+/// The `old_size` and `align` parameters are the parameters that were used to
+/// create the allocation referenced by `ptr`.
+#[allocator(reallocate)]
+pub fn reallocate(ptr: *mut u8, old_size: usize, size: usize, align: usize) -> *mut u8 {
+    ...
+}
+
+/// Resizes the allocation referenced by `ptr` to `size` bytes without moving
+/// it.
+///
+/// The new size of the allocation is returned. This must be at least
+/// `old_size`. The allocation must always remain valid.
+///
+/// Behavior is undefined if the requested size is 0 or the alignment is not a
```

Comment:
Should we go for "Behavior is undefined if the requested size is less than `old_size` or..."?
It might be worth spelling out explicitly whether `old_size` and `size` are the only legitimate return values or if the function can also return something inside that range.